flake-lang/flake-typescript.nix

```diff
@@ -38,181 +38,263 @@ pkgs:
 , # `testTools` are extra derivations to append to the `buildInputs` for
   # the tests (see the variable `test`)
   testTools ? [ ]
-, ...
 }:
-let
-  npmExtraDependenciesInfo =
-    pkgs.callPackage (import ./typescript/npm-extra-dependencies.nix) { } { inherit name npmExtraDependencies; };
-
-  projectNode2nix = pkgs.stdenv.mkDerivation {
-    name = "${name}-node2nix";
-    inherit src;
-    buildInputs = [ pkgs.node2nix nodejs npmExtraDependenciesInfo.npmLinkExtraDependencies ];
-    configurePhase =
-      ''
-        ${npmExtraDependenciesInfo.npmLinkExtraDependencies.name}
-      '';
-
-    buildPhase =
-      ''
-        NIX_NODE_ENV_FILE=./node-env.nix
-        NIX_NODE_PACKAGES_FILE=./node-packages.nix
-        NIX_COMPOSITION_FILE=./default.nix
-
-        # Make sure these files are generated with something like
-        # ```
-        # npm install --package-lock-only --lockfile-version 2
-        # ```
-        node2nix --input ./package.json --lock ./package-lock.json --development --node-env "$NIX_NODE_ENV_FILE" --output "$NIX_NODE_PACKAGES_FILE" --composition "$NIX_COMPOSITION_FILE"
-      '';
-
-    installPhase =
-      ''
-        mkdir -p $out
-        cp -r ./. $out
-      '';
-  };
-
-  projectNode2nixAttrs = import projectNode2nix { inherit nodejs pkgs; inherit (pkgs) system; };
-
-  # Important note:
-  # Inspection of the code suggests that the node_modules are put in 
-  # ```
-  # $out/lib/node_modules/${projectNode2nixAttrs.args.packageName}/node_modules
-  # ```
-  npmPackage = projectNode2nixAttrs.package.override
-    {
-      # Ensures that the node_modules has the extra linked dependencies when
-      # building it.
-      preRebuild =
-        ''
-          ${npmExtraDependenciesInfo.npmLinkExtraDependencies}/bin/${npmExtraDependenciesInfo.npmLinkExtraDependencies.name}
-        '';
-
-      # TODO(jaredponn): Wow this is horrible. `npm install` is broken for
-      # local dependencies on the filesystem. I think something like the
-      # following is problematic [this may not be an issue for later versions /
-      # lockfiles? needs more investigations]:
-      # - Suppose A is a tarball and depends on tarball B
-      # - Assume that we have a "sensible" `package.json` and
-      # `package-lock.json` with `./foo/A` and `./foo/B` installed
-      # - If we do _not_ have `node_modules`, and try to `npm install`, `npm
-      # install` will get confused and try to look in `node_modules/A/foo/B`
-      # which obviously doesn't exist so it errors.
-      # Apparently removing `package-lock.json` fixes this, so it can rebuild
-      # it from scratch I guess?
-      postRebuild =
-        ''
-          rm package-lock.json
-        '';
-    };
+pkgs.lib.makeExtensible
```

Comment:
This is the magic function which makes things "overlayable". I love it :^). The [source was surprisingly readable.](https://github.com/NixOS/nixpkgs/blob/master/lib/fixed-points.nix)

Most of the changes amount to:
- Moving all the above let bindings to the `__typescriptFlake__` attribute
- Putting the `let` bindings "in scope" by using `with self`.